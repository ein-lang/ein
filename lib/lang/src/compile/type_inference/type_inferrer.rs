use super::{
    super::{
        error::CompileError, reference_type_resolver::ReferenceTypeResolver,
        type_canonicalizer::TypeCanonicalizer, type_equality_checker::TypeEqualityChecker,
    },
    constraint_checker::ConstraintChecker,
    constraint_collector::ConstraintCollector,
    constraint_solver::ConstraintSolver,
    variable_substitutor::VariableSubstitutor,
};
use crate::{
    ast::*,
    types::{self, Type},
};
use std::sync::Arc;

pub struct TypeInferrer {
    reference_type_resolver: Arc<ReferenceTypeResolver>,
    type_equality_checker: Arc<TypeEqualityChecker>,
    type_canonicalizer: Arc<TypeCanonicalizer>,
    constraint_collector: ConstraintCollector,
    constraint_solver: Arc<ConstraintSolver>,
}

impl TypeInferrer {
    pub fn new(
        reference_type_resolver: Arc<ReferenceTypeResolver>,
        type_equality_checker: Arc<TypeEqualityChecker>,
        type_canonicalizer: Arc<TypeCanonicalizer>,
        constraint_collector: ConstraintCollector,
        constraint_solver: Arc<ConstraintSolver>,
    ) -> Self {
        Self {
            reference_type_resolver,
            type_equality_checker,
            type_canonicalizer,
            constraint_collector,
            constraint_solver,
        }
    }

    pub fn infer(self, module: &Module) -> Result<Module, CompileError> {
        let module = module.transform_types(&mut |type_| -> Result<_, CompileError> {
            Ok(match type_ {
                Type::Unknown(unknown) => {
                    types::Variable::new(unknown.source_information().clone()).into()
                }
                _ => type_.clone(),
            })
        })?;

        let (solved_subsumption_set, mut checked_subsumption_set) =
            self.constraint_collector.collect(&module)?;

        let substitutions = self
            .constraint_solver
            .solve(solved_subsumption_set, &mut checked_subsumption_set)?;

        let substitutor = VariableSubstitutor::new(self.type_canonicalizer.clone(), substitutions);

        let checker = ConstraintChecker::new(
            substitutor.clone(),
            self.reference_type_resolver.clone(),
            self.type_equality_checker,
        );

        checker.check(checked_subsumption_set)?;

        module.transform_types(&mut |type_| substitutor.substitute(type_))
    }
}
