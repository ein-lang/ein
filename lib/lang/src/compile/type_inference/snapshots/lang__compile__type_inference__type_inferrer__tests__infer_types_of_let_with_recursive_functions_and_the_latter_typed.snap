---
source: lib/lang/src/compile/type_inference/type_inferrer.rs
expression: "infer_types(&Module::from_definitions(vec![VariableDefinition ::\n                                           new(\"x\", LetRecursive ::\n                                               new(vec !\n                                                   [FunctionDefinition ::\n                                                    new(\"f\", vec !\n                                                        [\"x\" . into()],\n                                                        Application ::\n                                                        new(Variable ::\n                                                            new(\"g\",\n                                                                SourceInformation\n                                                                :: dummy()),\n                                                            Variable ::\n                                                            new(\"x\",\n                                                                SourceInformation\n                                                                :: dummy()),\n                                                            SourceInformation\n                                                            :: dummy(),),\n                                                        types :: Unknown ::\n                                                        new(SourceInformation\n                                                            :: dummy()),\n                                                        SourceInformation ::\n                                                        dummy()) . into(),\n                                                    FunctionDefinition ::\n                                                    new(\"g\", vec !\n                                                        [\"x\" . into()],\n                                                        Application ::\n                                                        new(Variable ::\n                                                            new(\"f\",\n                                                                SourceInformation\n                                                                :: dummy()),\n                                                            Variable ::\n                                                            new(\"x\",\n                                                                SourceInformation\n                                                                :: dummy()),\n                                                            SourceInformation\n                                                            :: dummy(),),\n                                                        types :: Function ::\n                                                        new(types :: Number ::\n                                                            new(SourceInformation\n                                                                :: dummy()),\n                                                            types :: Number ::\n                                                            new(SourceInformation\n                                                                :: dummy()),\n                                                            SourceInformation\n                                                            :: dummy(),),\n                                                        SourceInformation ::\n                                                        dummy()) . into()],\n                                                   Application ::\n                                                   new(Variable ::\n                                                       new(\"f\",\n                                                           SourceInformation\n                                                           :: dummy()), Number\n                                                       ::\n                                                       new(42.0,\n                                                           SourceInformation\n                                                           :: dummy()),\n                                                       SourceInformation ::\n                                                       dummy()),\n                                                   SourceInformation ::\n                                                   dummy(),), types :: Number\n                                               ::\n                                               new(SourceInformation ::\n                                                   dummy()), SourceInformation\n                                               :: dummy(),) . into()]))"
---
Ok(
    Module {
        path: ModulePath {
            package: Package {
                name: "",
                version: "",
            },
            components: [],
        },
        type_definitions: [],
        definitions: [
            VariableDefinition(
                VariableDefinition {
                    name: "x",
                    body: LetRecursive(
                        LetRecursive {
                            definitions: [
                                FunctionDefinition {
                                    name: "f",
                                    arguments: [
                                        "x",
                                    ],
                                    body: Application(
                                        Application {
                                            function: Variable(
                                                Variable {
                                                    name: "g",
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            argument: Variable(
                                                Variable {
                                                    name: "x",
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    type_: Function(
                                        Function {
                                            argument: Number(
                                                Number {
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            result: Number(
                                                Number {
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                                FunctionDefinition {
                                    name: "g",
                                    arguments: [
                                        "x",
                                    ],
                                    body: Application(
                                        Application {
                                            function: Variable(
                                                Variable {
                                                    name: "f",
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            argument: Variable(
                                                Variable {
                                                    name: "x",
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    type_: Function(
                                        Function {
                                            argument: Number(
                                                Number {
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            result: Number(
                                                Number {
                                                    source_information: SourceInformation {
                                                        source_name: "",
                                                        location: Location {
                                                            line_number: 0,
                                                            column_number: 0,
                                                        },
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ],
                            expression: Application(
                                Application {
                                    function: Variable(
                                        Variable {
                                            name: "f",
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    argument: Number(
                                        Number {
                                            value: 42.0,
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    type_: Number(
                        Number {
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
        ],
        export: Export {
            names: {},
        },
        imports: [],
    },
)
